### PR TITLE
feat(agent-codex): mature Codex plugin to match Claude Code support

### LINKED ISSUE #147: fix: Codex plugin uses wrong permission flag for unattended mode
## Summary

The Codex agent plugin uses `--approval-mode full-auto` when `permissions: skip` is configured. This is not the correct Codex CLI flag, so Codex doesn't actually run in unattended mode — it still prompts for approvals.

## Current State

```typescript
// agent-codex/src/index.ts — getLaunchCommand()
if (config.permissions === "skip") {
  parts.push("--approval-mode", "full-auto"); // ❌ Wrong flag
}
```

## Fix

Replace with the correct flag, verified against [dmux](https://github.com/standardagents/dmux)'s production-tested Codex integration:

```typescript
if (config.permissions === "skip") {
  parts.push("--dangerously-bypass-approvals-and-sandbox"); // ✅ Correct
}
```

## Files Changed

- `packages/plugins/agent-codex/src/index.ts` — `getLaunchCommand()`

## Related

- #17 — Codex: implement getSessionInfo and richer activity detection

### LINKED ISSUE #148: fix: Codex detectActivity() is a stub — returns active for any output
## Summary

The Codex agent plugin's `detectActivity()` returns `"active"` for any non-empty terminal output. This means the orchestrator can never distinguish between an idle Codex (waiting for input), a Codex awaiting approval, or a Codex actively working. Stuck detection and needs-input detection don't work.

## Current State

```typescript
detectActivity(terminalOutput: string): ActivityState {
  if (!terminalOutput.trim()) return "idle";
  // Codex doesn't have rich terminal output patterns yet
  return "active"; // ❌ Always returns active
}
```

## Fix

Implement terminal pattern matching following the same structure as the Claude Code plugin's `classifyTerminalOutput()`, adapted with Codex-specific patterns from [dmux](https://github.com/standardagents/dmux):

| Pattern | State | Source |
|---------|-------|--------|
| Last line is bare prompt (`>`, `$`, `#`) | `idle` | dmux agentLaunch.ts |
| Last 5 lines contain `approval required` or `(y)es (n)o` | `waiting_input` | dmux PaneAnalyzer.ts |
| `(esc to interrupt)` visible | `active` | dmux PaneWorker.ts |
| Progress spinners `✶⏺✽⏳` + `"ing"` words | `active` | dmux PaneWorker.ts |
| Fallback | `active` | — |

## Files Changed

- `packages/plugins/agent-codex/src/index.ts` — `detectActivity()`

## Related

- #17 — Codex: implement getSessionInfo and richer activity detection
- #16 — Claude Code: detect waiting_input and blocked states via hooks

### LINKED ISSUE #149: feat: Codex setupWorkspaceHooks and postLaunchSetup with PATH-based shell wrappers
## Summary

The Codex agent plugin does not implement `setupWorkspaceHooks()` or `postLaunchSetup()`. This means:

- No metadata auto-updates when Codex runs `gh pr create`, `git checkout -b`, or `gh pr merge`
- Dashboard is blind to PRs created by Codex agents
- CI monitoring, review detection, and notifications don't work for Codex sessions

Claude Code solves this with its built-in PostToolUse hook system. Codex has no equivalent, so we need an alternative.

### PATH-based shell wrappers

Install thin wrapper scripts for `gh` and `git` at `~/.ao/bin/` and prepend that directory to `PATH` in `getEnvironment()`. When Codex runs `gh` or `git`, the shell finds our wrapper first.

Each wrapper:
1. Strips `~/.ao/bin` from PATH to find the real binary (no infinite recursion)
2. Runs the real command, passing through all args and exit code
3. On success, parses output for metadata-relevant events
4. Updates the session metadata file using `$AO_SESSION` and `$AO_DATA_DIR` env vars

| Wrapper | Intercepts | Updates |
|---------|-----------|---------|
| `~/.ao/bin/gh` | `gh pr create`, `gh pr merge` | `pr=<url>`, `status=pr_open`, `status=merged` |
| `~/.ao/bin/git` | `git checkout -b`, `git switch -c` | `branch=<name>` |
| `~/.ao/bin/ao-metadata-helper.sh` | — | Shared `update_ao_metadata()` function |

### AGENTS.md (secondary signal)

Append an ao section to workspace `AGENTS.md` as defense-in-depth. The wrappers are the primary mechanism; AGENTS.md reinforces the intent if wrappers are somehow bypassed.

### getEnvironment() PATH prepend

```typescript
getEnvironment(config: AgentLaunchConfig): Record<string, string> {
  const env: Record<string, string> = {};
  // ... existing env vars ...
  env["PATH"] = `${AO_BIN_DIR}:${process.env["PATH"] ?? "/usr/bin:/bin"}`;
  return env;
}
```

## Why PATH wrappers?

This is the same pattern used by nvm, rbenv, and pyenv. It works because Codex executes commands through the shell (tmux pane), so PATH resolution applies. The wrappers are transparent — Codex never knows they're there.

## Files Changed

- `packages/plugins/agent-codex/src/index.ts` — `setupWorkspaceHooks()`, `postLaunchSetup()`, `getEnvironment()`

## Related

- #17 — Codex: implement getSessionInfo and richer activity detection
- #147 — Codex permission flag fix
- #148 — Codex detectActivity fix

### LINKED ISSUE #150: feat: branch-based PR auto-detection fallback in lifecycle manager
## Summary

The lifecycle manager's `determineStatus()` gates all PR checks on `session.pr` being set:

```typescript
if (session.pr && scm) {
  // CI monitoring, review detection, mergeability — ALL skipped if session.pr is null
}
```

For Claude Code, `session.pr` is auto-populated via PostToolUse hooks. But agents without auto-hook systems (Codex, Aider, OpenCode) may never write `pr=<url>` to metadata, causing the entire PR state machine to be skipped — no CI monitoring, no review detection, no notifications.

## Current State

- `session.pr` is null → lifecycle manager skips PR checks entirely
- Agent created a PR but metadata was never updated
- Dashboard shows no PR card, no CI status, no reviews
- Notifications for CI failures and review comments never fire

## Proposed Implementation

Add a branch-based PR detection step before the PR state check. The SCM interface already has `detectPR(session, project)` which queries `gh pr list --head <branch>`:

```typescript
// Before checking PR state, auto-detect PR by branch if metadata.pr is missing
if (!session.pr && scm && session.branch) {
  const detectedPR = await scm.detectPR(session, project);
  if (detectedPR) {
    session.pr = detectedPR;
    // Persist to metadata so subsequent polls don't need to re-query
    updateMetadata(sessionsDir, session.id, {
      pr: detectedPR.url,
      status: "pr_open",
    });
  }
}
```

Once detected, the PR URL is persisted to metadata so future poll cycles don't re-query GitHub.

## Impact

This is agent-agnostic — benefits every agent that lacks auto-hook systems:
- **Codex**: no PostToolUse hooks
- **Aider**: no hook system
- **OpenCode**: no hook system
- **Future agents**: automatic support

## Files Changed

- `packages/core/src/lifecycle-manager.ts` — `determineStatus()`

## Related

- #149 — Codex PATH-based shell wrappers (primary metadata mechanism)
- #17 — Codex: implement getSessionInfo and richer activity detection

### PR DESCRIPTION
**Pull request by suraj-markup**
## Summary

Brings the Codex agent plugin to feature parity with Claude Code by fixing four key gaps identified via cross-referencing with the [dmux](https://github.com/anthropics/dmux) codebase (production-tested Codex orchestration patterns).

### Code changes

**`packages/plugins/agent-codex/src/index.ts`** — Codex agent plugin (main changes)

- **Permission flag fix**: Changed `--approval-mode full-auto` → `--dangerously-bypass-approvals-and-sandbox` to match the actual Codex CLI flag for unattended mode
- **Terminal-based activity detection**: Implemented `detectActivity()` with Codex-specific terminal patterns:
  - `idle` — bare prompt (`>`, `$`, `#`), empty output
  - `waiting_input` — approval prompts (`approval required`, `(y)es / (n)o`)
  - `active` — `(esc to interrupt)`, spinner symbols (`✶⏺✽⏳`) with `-ing` words
- **PATH-based shell wrappers**: Installs `gh` and `git` wrapper scripts to `~/.ao/bin/` that intercept commands and auto-update session metadata (same pattern as nvm/rbenv/pyenv). Wrappers:
  - `gh` wrapper — captures `gh pr create` / `gh pr merge` output, extracts PR URL, updates `pr` and `status` metadata. All other commands use `exec` for transparent passthrough (no stderr merging)
  - `git` wrapper — captures `git checkout -b` / `git switch -c`, updates `branch` metadata
  - `ao-metadata-helper.sh` — shared helper providing `update_ao_metadata <key> <value>`
- **`getEnvironment()`** — Prepends `~/.ao/bin` to `PATH` so wrappers intercept commands without agent awareness
- **`setupWorkspaceHooks()`** / **`postLaunchSetup()`** — Writes wrappers + appends AGENTS.md section as secondary signal
- **`getActivityState()`** — Returns `{ state: "exited" }` when process is dead, `null` (unknown) when running (Codex lacks per-session tracking)

**`packages/core/src/lifecycle-manager.ts`** — Branch-based PR fallback

- Added step 3 in `determineStatus()`: when `metadata.pr` is missing, calls `scm.detectPR(session, project)` which runs `gh pr list --head <branch>` to auto-discover PRs created by any agent (not just Claude Code which has PostToolUse hooks)
- Gracefully handles SCM failures — retries on next poll cycle

**`packages/cli/src/commands/spawn.ts`** + **`packages/core/src/types.ts`** + **`packages/core/src/session-manager.ts`** — `--agent` CLI override

- Added `--agent <name>` option to `ao spawn` to override the default agent plugin at spawn time
- Allows starting Codex (or any agent) sessions without changing config:
  ```bash
  ao spawn ao --agent codex           # interactive mode
  ao spawn ao --agent codex INT-1234  # with an issue
  ```
- Added `agent?: string` to `SessionSpawnConfig` interface
- Session manager resolves the override agent from the plugin registry before launch

### Security hardening (addressing BugBot review)

- **Sed metacharacter escaping**: Values passed to `sed` are escaped for `&`, `|`, `\` to prevent metadata corruption from branch names like `feat&fix`
- **Fixed-string grep**: PATH cleaning uses `grep -Fxv` instead of regex `grep -v` so `.ao` is matched literally
- **Empty PATH element filtering**: `grep .` removes empty PATH entries to prevent CWD search
- **Newline stripping**: Values sanitized with `tr -d '\n'` to prevent metadata line injection
- **Path validation**: `update_ao_metadata()` validates `AO_SESSION` has no path separators or `..` traversal, `AO_DATA_DIR` is under `~/.ao/`, `~/.agent-orchestrator/`, or `/tmp/`, and resolves symlinks to verify the target stays within the data directory
- **Atomic wrapper writes**: Wrapper files written via temp file + rename to prevent concurrent sessions from reading partially written scripts. Version marker written last for crash-safe idempotency
- **Pipe instead of process substitution**: gh wrapper uses `2>&1 | tee` instead of `> >(tee)` to avoid race condition where tee may not finish writing before output is read

### Why PATH-based wrappers?

Claude Code has built-in `PostToolUse` hooks that fire after every tool invocation, making metadata updates automatic. Codex has no equivalent hook system. Instead of relying on AGENTS.md instructions (which LLMs may ignore), we intercept the actual shell commands via PATH precedence — the same battle-tested pattern used by nvm, rbenv, and pyenv. The wrappers:
1. Remove `~/.ao/bin` from PATH to find the real binary (no infinite recursion)
2. Run the real command, capturing output only for PR/branch operations
3. Parse output for metadata (PR URLs, branch names)
4. Update the session metadata file via env vars (`AO_SESSION`, `AO_DATA_DIR`)

### Test coverage

- **85 tests pass** in `packages/plugins/agent-codex/src/index.test.ts` (up from 27)
  - Updated 3 existing tests for new permission flag
  - Added 58 new tests replicating Claude Code test patterns:
    - `detectActivity` edge cases (idle/active/waiting_input, regression tests for history vs current state)
    - `isProcessRunning` boundaries (empty panes, name boundary matching, string PIDs)
    - `setupWorkspaceHooks` file I/O behavior (atomic writes, version skipping, AGENTS.md idempotency)
    - Shell wrapper content verification (sed escaping, grep -Fxv, exec passthrough, PR parsing)
- **245 tests pass** in `packages/core/`
- **All 22 packages typecheck clean**

## Closes

- Closes #147 — Codex plugin uses wrong permission flag
- Closes #148 — Codex detectActivity() is a stub
- Closes #149 — Codex setupWorkspaceHooks and postLaunchSetup with PATH-based shell wrappers
- Closes #150 — Branch-based PR auto-detection fallback in lifecycle manager

## Test plan

- [x] `npx vitest run` in `packages/plugins/agent-codex/` — 85 tests pass
- [x] `npx vitest run` in `packages/core/` — 245 tests pass
- [x] `pnpm typecheck` — all 22 packages pass
- [x] `pnpm build` — core + CLI build clean
- [ ] Manual: `ao spawn ao --agent codex` — verify Codex starts in interactive mode
- [ ] Manual: `ao spawn ao --agent codex INT-1234` — verify Codex starts with issue prompt
- [ ] Manual: Verify `~/.ao/bin/` wrappers are created on session start
- [ ] Manual: Run `gh pr create` inside Codex session, verify metadata file updates with PR URL
- [ ] Manual: Run `git checkout -b feat/test` inside Codex session, verify branch metadata updates
- [ ] Manual: Verify lifecycle manager picks up PRs for Codex sessions via branch detection

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by suraj-markup** [comment]
Yes, sure.

File: packages/plugins/agent-codex/src/index.ts [hunk 1]
```diff
@@ -25,6 +33,202 @@ export const manifest = {
...
+
+  [[ -z "\$ao_dir" || -z "\$ao_session" ]] && return 0
+
+  local metadata_file="\$ao_dir/\$ao_session"
+  [[ -f "\$metadata_file" ]] || return 0
+
+  local temp_file="\${metadata_file}.tmp.\$\$"
+
+  if grep -q "^\${key}=" "\$metadata_file" 2>/dev/null; then
+    sed "s|^\${key}=.*|\${key}=\${value}|" "\$metadata_file" > "\$temp_file"
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts]: Fixed in 77bbade — added sed metacharacter escaping. The `update_ao_metadata` function now escapes `&`, `|`, and `\` in values before passing to `sed`, while the `printf` append path (which doesn't go through sed) continues to use the raw value.

File: packages/plugins/agent-codex/src/index.ts [hunk 2]
```diff
@@ -25,6 +33,202 @@ export const manifest = {
...
+/**
+ * gh wrapper — intercepts `gh pr create` and `gh pr merge` to auto-update
+ * session metadata. All other commands pass through transparently.
+ */
+const GH_WRAPPER = `#!/usr/bin/env bash
+# ao gh wrapper — auto-updates session metadata on PR operations
+
+# Find real gh by removing our wrapper directory from PATH
+ao_bin_dir="\$(cd "\$(dirname "\$0")" && pwd)"
+clean_path="\$(echo "\$PATH" | tr ':' '\\n' | grep -v "^\$ao_bin_dir\$" | tr '\\n' ':')"
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts]: Fixed in 77bbade — changed `grep -v "^$ao_bin_dir$"` to `grep -Fxv "$ao_bin_dir"` in both the `gh` and `git` wrappers. `-F` treats the pattern as a fixed string (so `.ao` is literal), `-x` ensures whole-line matching.

File: packages/plugins/agent-codex/src/index.ts [hunk 3]
```diff
@@ -25,6 +33,205 @@ export const manifest = {
...
+
+# Source the metadata helper
+source "\$ao_bin_dir/ao-metadata-helper.sh" 2>/dev/null || true
+
+# Capture output while streaming it to the terminal
+tmpout="\$(mktemp)"
+trap 'rm -f "\$tmpout"' EXIT
+
+"\$real_gh" "\$@" 2>&1 | tee "\$tmpout"
+exit_code=\${PIPESTATUS[0]}
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts]: Fixed in ec41ff2 — the gh wrapper now only captures output for `pr/create` and `pr/merge` via a `case` statement. All other commands use `exec "$real_gh" "$@"` for transparent passthrough with no stream merging.

File: packages/plugins/agent-codex/src/index.ts [hunk 4]
```diff
@@ -25,6 +33,211 @@ export const manifest = {
...
+
+  if grep -q "^\${key}=" "\$metadata_file" 2>/dev/null; then
+    sed "s|^\${key}=.*|\${key}=\${escaped_value}|" "\$metadata_file" > "\$temp_file"
+  else
+    cp "\$metadata_file" "\$temp_file"
+    printf '%s=%s\\n' "\$key" "\$value" >> "\$temp_file"
+  fi
+
+  mv "\$temp_file" "\$metadata_file"
+}
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:93]: Fixed in 4ca202b — `update_ao_metadata()` now validates:
1. `AO_SESSION` contains no path separators (`/`) or traversal (`..`)
2. `AO_DATA_DIR` is an absolute path under `~/.ao/` or `/tmp/`
3. After constructing the path, resolves symlinks via `pwd -P` and verifies the target directory is still within `AO_DATA_DIR`

All validations bail silently (`return 0`) to avoid disrupting the wrapped command.

File: packages/plugins/agent-codex/src/index.ts [hunk 5]
```diff
@@ -25,6 +33,211 @@ export const manifest = {
...
+  if (needsUpdate) {
+    await writeFile(join(AO_BIN_DIR, "gh"), GH_WRAPPER, "utf-8");
+    await chmod(join(AO_BIN_DIR, "gh"), 0o755);
+
+    await writeFile(join(AO_BIN_DIR, "git"), GIT_WRAPPER, "utf-8");
+    await chmod(join(AO_BIN_DIR, "git"), 0o755);
+
+    await writeFile(markerPath, currentVersion, "utf-8");
+  }
+
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:255]: Fixed in 4ca202b — wrapper files are now written atomically via `atomicWriteFile()` which writes to a temp file (random suffix) then renames. The version marker is written last, so a crash mid-install leaves the marker stale and triggers a safe redo on the next invocation.

File: packages/plugins/agent-codex/src/index.ts [hunk 6]
```diff
@@ -25,6 +33,238 @@ export const manifest = {
...
+ * Atomically write a file by writing to a temp file in the same directory,
+ * then renaming. This prevents concurrent sessions from reading partially
+ * written wrapper scripts.
+ */
+async function atomicWriteFile(filePath: string, content: string, mode: number): Promise<void> {
+  const suffix = randomBytes(6).toString("hex");
+  const tmpPath = `${filePath}.tmp.${suffix}`;
+  await writeFile(tmpPath, content, { encoding: "utf-8", mode });
+  await rename(tmpPath, filePath);
+}
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:224]: Won't fix — this project targets macOS and Linux only (tmux runtime, bash wrappers, POSIX shell scripts). Windows is not a supported platform. On POSIX systems, `rename(2)` atomically replaces the destination file if it already exists, which is exactly the behavior we rely on.

File: packages/plugins/agent-codex/src/index.ts [hunk 7]
```diff
@@ -25,6 +33,238 @@ export const manifest = {
...
+
+  # Escape sed metacharacters in value (& expands to matched text, | breaks delimiter)
+  local escaped_value="\$(printf '%s' "\$value" | sed 's/[&|\\\\]/\\\\&/g')"
+
+  if grep -q "^\${key}=" "\$metadata_file" 2>/dev/null; then
+    sed "s|^\${key}=.*|\${key}=\${escaped_value}|" "\$metadata_file" > "\$temp_file"
+  else
+    cp "\$metadata_file" "\$temp_file"
+    printf '%s=%s\\n' "\$key" "\$value" >> "\$temp_file"
+  fi
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:90]: Won't fix — the keys passed to `update_ao_metadata` are hardcoded string literals in the wrappers themselves (`pr`, `status`, `branch`), not user-supplied input. The wrappers call `update_ao_metadata branch "$3"` where only the value comes from external input (already escaped for sed in our earlier fix). A crafted key would require modifying the wrapper scripts themselves, at which point the attacker already has arbitrary write access.

File: packages/plugins/agent-codex/src/index.ts [hunk 8]
```diff
@@ -25,6 +33,238 @@ export const manifest = {
...
+  # Validate: session name must not contain path separators or traversal
+  case "\$ao_session" in
+    */* | *..*) return 0 ;;
+  esac
+
+  # Validate: ao_dir must be an absolute path under ~/.ao or /tmp
+  case "\$ao_dir" in
+    "\$HOME"/.ao/* | /tmp/*) ;;
+    *) return 0 ;;
+  esac
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:65]: Good catch — fixed in e24f34a. The allowlist now includes `$HOME/.agent-orchestrator/*` alongside `$HOME/.ao/*` and `/tmp/*`, matching the actual data directory used by core (`~/.agent-orchestrator/{hash}-{project}/sessions/`).

File: packages/plugins/agent-codex/src/index.ts [hunk 9]
```diff
@@ -3,17 +3,25 @@ import {
...
-  type ActivityDetection,
   type ActivityState,
+  type ActivityDetection,
   type PluginModule,
   type RuntimeHandle,
   type Session,
+  type WorkspaceHooksConfig,
 } from "@composio/ao-core";
 import { execFile } from "node:child_process";
+import { writeFile, mkdir, chmod, readFile, rename } from "node:fs/promises";
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts]: Fixed in e24f34a — removed the unused `chmod` import. With the switch to `atomicWriteFile()`, file permissions are now set via the `mode` option passed directly to `writeFile()`, so the separate `chmod` call is no longer needed.

File: packages/plugins/agent-codex/src/index.ts [hunk 10]
```diff
@@ -25,6 +33,238 @@ export const manifest = {
...
+case "\$1/\$2" in
+  pr/create|pr/merge)
+    tmpout="\$(mktemp)"
+    trap 'rm -f "\$tmpout"' EXIT
+
+    "\$real_gh" "\$@" > >( tee "\$tmpout" ) 2>&1
+    exit_code=\${PIPESTATUS[0]}
+
+    if [[ \$exit_code -eq 0 ]]; then
+      output="\$(cat "\$tmpout")"
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:128]: Fixed in 90798ac — replaced process substitution `> >(tee "$tmpout")` with a proper pipe `2>&1 | tee "$tmpout"`. Bash waits for all pipe segments to complete, so `PIPESTATUS[0]` correctly captures the exit code and `cat "$tmpout"` reads the full output.

File: packages/core/src/lifecycle-manager.ts [hunk 11]
```diff
@@ -229,7 +229,27 @@ export function createLifecycleManager(deps: LifecycleManagerDeps): LifecycleMan
...
+        const detectedPR = await scm.detectPR(session, project);
+        if (detectedPR) {
+          session.pr = detectedPR;
+          // Persist to metadata so subsequent polls don't need to re-query
+          const sessionsDir = getSessionsDir(config.configPath, project.path);
+          updateMetadata(sessionsDir, session.id, {
+            pr: detectedPR.url,
+            status: "pr_open",
+          });
+        }
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/core/src/lifecycle-manager.ts:246]: Fixed in f5c609f — removed the premature `status: "pr_open"` write from the auto-detect step. Now only the `pr` URL is persisted there; the correct status (merged, ci_failed, changes_requested, etc.) is determined by step 4 on the same poll cycle, avoiding any one-cycle metadata mismatch.

File: packages/plugins/agent-codex/src/index.ts [hunk 12]
```diff
@@ -25,6 +33,238 @@ export const manifest = {
...
+  if grep -q "^\${key}=" "\$metadata_file" 2>/dev/null; then
+    sed "s|^\${key}=.*|\${key}=\${escaped_value}|" "\$metadata_file" > "\$temp_file"
+  else
+    cp "\$metadata_file" "\$temp_file"
+    printf '%s=%s\\n' "\$key" "\$value" >> "\$temp_file"
+  fi
+
+  mv "\$temp_file" "\$metadata_file"
+}
+`;
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:94]: Fixed in 4d957aa — values are now sanitized with `tr -d '\n'` before both the sed replacement and printf append paths. The new `clean_value` variable strips newlines early, preventing metadata line injection regardless of which code path is taken.

File: packages/plugins/agent-codex/src/index.ts [hunk 13]
```diff
@@ -25,6 +33,238 @@ export const manifest = {
...
+          pr_url="\$(echo "\$output" | grep -Eo 'https://github\\.com/[^/]+/[^/]+/pull/[0-9]+' | head -1)"
+          if [[ -n "\$pr_url" ]]; then
+            update_ao_metadata pr "\$pr_url"
+            update_ao_metadata status pr_open
+          fi
+          ;;
+        pr/merge)
+          update_ao_metadata status merged
+          ;;
+      esac
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:140]: By design — the wrapper writes `status pr_open` as an immediate signal so the dashboard reflects the new PR without waiting for the next lifecycle poll cycle. The lifecycle manager overwrites this with the correct SCM-derived status (merged, ci_failed, changes_requested, etc.) on the very next poll. The brief one-cycle desync is harmless and the tradeoff is worth it for instant feedback.

File: packages/plugins/agent-codex/src/index.ts [hunk 14]
```diff
@@ -25,6 +33,241 @@ export const manifest = {
...
+ * session metadata. All other commands pass through transparently.
+ */
+const GH_WRAPPER = `#!/usr/bin/env bash
+# ao gh wrapper — auto-updates session metadata on PR operations
+
+# Find real gh by removing our wrapper directory from PATH
+ao_bin_dir="\$(cd "\$(dirname "\$0")" && pwd)"
+clean_path="\$(echo "\$PATH" | tr ':' '\\n' | grep -Fxv "\$ao_bin_dir" | tr '\\n' ':')"
+clean_path="\${clean_path%:}"
+real_gh="\$(PATH="\$clean_path" command -v gh 2>/dev/null)"
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:107]: Fixed in e410079 — added `grep .` to the PATH cleaning pipeline in both gh and git wrappers. This filters out empty lines (from leading `:`, trailing `:`, or `::` in PATH) before converting back to colon-separated format, preventing the shell from searching the current working directory.

File: packages/plugins/agent-codex/src/index.ts [hunk 15]
```diff
@@ -25,6 +33,241 @@ export const manifest = {
...
+    // File doesn't exist yet
+  }
+
+  if (!existing.includes("Agent Orchestrator (ao) Session")) {
+    const content = existing
+      ? existing.trimEnd() + "\n" + AO_AGENTS_MD_SECTION
+      : AO_AGENTS_MD_SECTION.trimStart();
+    await writeFile(agentsMdPath, content, "utf-8");
+  }
+}
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:271]: Won't fix — the project currently only supports local runtimes (tmux, process). Container and remote runtimes (Docker, k8s) don't exist yet. When they're added, all plugins that touch the host filesystem (not just Codex wrappers) will need container-aware path resolution — that's a cross-cutting concern for the runtime abstraction layer, not something to solve in a single agent plugin. The wrappers gracefully degrade: if `~/.ao/bin` doesn't exist in the agent's environment, the wrappers simply aren't found in PATH, and the real `gh`/`git` are used directly (metadata just won't auto-update, which the lifecycle manager's branch-based PR detection covers as a fallback).

File: packages/core/src/session-manager.ts [hunk 16]
```diff
@@ -322,6 +322,16 @@ export function createSessionManager(deps: SessionManagerDeps): SessionManager {
...
     }
+
+    // Allow --agent override to swap the agent plugin for this session
+    if (spawnConfig.agent) {
+      const overrideAgent = registry.get<Agent>("agent", spawnConfig.agent);
+      if (!overrideAgent) {
+        throw new Error(`Agent plugin '${spawnConfig.agent}' not found`);
+      }
+      plugins.agent = overrideAgent;
+    }
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/core/src/session-manager.ts:333]: Fixed in 46b0d1a — the agent name is now persisted in session metadata and read back by both the lifecycle manager and session enrichment paths.

Changes:
- Added `agent` field to `SessionMetadata` type and `writeMetadata`
- Spawn now writes `agent: plugins.agent.name` to metadata
- `resolvePlugins()` accepts optional `agentOverride` param, used by list/get/restore
- Lifecycle manager reads `session.metadata["agent"]` instead of always using project default
- Added 2 tests verifying agent name persistence with and without override

File: packages/core/src/metadata.ts [hunk 17]
```diff
@@ -139,6 +139,7 @@ export function writeMetadata(
   if (metadata.pr) data["pr"] = metadata.pr;
   if (metadata.summary) data["summary"] = metadata.summary;
   if (metadata.project) data["project"] = metadata.project;
+  if (metadata.agent) data["agent"] = metadata.agent;
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/core/src/metadata.ts:143]: Fixed in 89fc336 — added `agent: raw["agent"]` to the `readMetadata()` return mapping. The typed `SessionMetadata` interface already had the `agent` field (added in 46b0d1a), but the parsing function was not mapping it from the raw key-value file.

Added a test (`readMetadata returns agent field (typed SessionMetadata)`) that spawns with `--agent codex`, then verifies `readMetadata()` returns `meta.agent === "codex"`.

File: packages/plugins/agent-codex/src/index.ts [hunk 18]
```diff
@@ -38,7 +283,7 @@ function createCodexAgent(): Agent {
       const parts: string[] = ["codex"];
 
       if (config.permissions === "skip") {
-        parts.push("--approval-mode", "full-auto");
+        parts.push("--dangerously-bypass-approvals-and-sandbox");
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts]: By design — this is the same pattern used by the Claude Code plugin (`--append-system-prompt "$(cat ...)"` at line 603 of agent-claude-code/src/index.ts). The command runs inside a bash shell (via tmux `send-keys`), so `$(cat ...)` expands correctly at launch time.

Neither Codex CLI nor Claude Code CLI support a `--system-prompt-file` flag, so shell command substitution is the intended mechanism. The path is already sanitized via `shellEscape()` to prevent injection.

File: packages/plugins/agent-codex/src/index.ts [hunk 19]
```diff
@@ -67,21 +312,43 @@ function createCodexAgent(): Agent {
...
+      const tail = lines.slice(-5).join("\n");
+      if (/approval required/i.test(tail)) return "waiting_input";
+      if (/\(y\)es.*\(n\)o/i.test(tail)) return "waiting_input";
+
+      // "(esc to interrupt)" appears while Codex is actively running a task
+      if (/\(esc to interrupt\)/i.test(terminalOutput)) return "active";
+
+      // Progress spinner symbols + words ending in "ing" = actively working
+      if (/[✶⏺✽⏳]/.test(terminalOutput) && /\w+ing\b/.test(terminalOutput)) return "active";
+
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts]: Fixed in 584ca1d — removed the redundant `return "active"` checks. The default already returns `"active"`, so the explicit pattern checks for `(esc to interrupt)` and spinner symbols were dead code. Added a comment mirroring the Claude Code plugin's approach.

Also in the same commit: replaced the nonexistent `--system-prompt` flag with correct Codex config overrides (`-c developer_instructions=...` and `-c model_instructions_file=...`), and switched `--dangerously-bypass-approvals-and-sandbox` to `--full-auto` (keeps workspace sandbox, safer default for automation).

File: packages/plugins/agent-codex/src/index.ts [hunk 20]
```diff
@@ -38,17 +283,19 @@ function createCodexAgent(): Agent {
       const parts: string[] = ["codex"];
 
       if (config.permissions === "skip") {
-        parts.push("--approval-mode", "full-auto");
+        parts.push("--full-auto");
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:286]: Fixed in 0a5fea1 — updated all test expectations to use `--full-auto` instead of `--dangerously-bypass-approvals-and-sandbox`. All 85 codex plugin tests pass.

File: packages/plugins/agent-codex/src/index.ts [hunk 21]
```diff
@@ -67,21 +314,39 @@ function createCodexAgent(): Agent {
       if (config.issueId) {
         env["AO_ISSUE_ID"] = config.issueId;
       }
+
+      // Prepend ~/.ao/bin to PATH so our gh/git wrappers intercept commands.
+      // The wrappers strip this directory from PATH before calling the real
+      // binary, so there's no infinite recursion.
+      env["PATH"] = `${AO_BIN_DIR}:${process.env["PATH"] ?? "/usr/bin:/bin"}`;
+
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.ts:322]: Not a bug — `AO_DATA_DIR` and `AO_SESSION` are set by the session manager in core (`session-manager.ts:480-481`), not by the agent plugin. The plugin's `getEnvironment()` only returns agent-specific env vars (like `AO_SESSION_ID`, `PATH`). Core merges these with its own env vars before passing to the runtime. The Claude Code plugin follows the same pattern — its `getEnvironment()` also doesn't set `AO_DATA_DIR`/`AO_SESSION`.

File: packages/plugins/agent-codex/src/index.test.ts [hunk 22]
```diff
@@ -133,18 +167,46 @@ describe("getLaunchCommand", () => {
...
+      makeLaunchConfig({ systemPromptFile: "/tmp/prompt.md", systemPrompt: "Ignored" }),
+    );
+    expect(cmd).toContain("$(cat '/tmp/prompt.md')");
+    expect(cmd).not.toContain("'Ignored'");
+  });
+
+  it("includes --system-prompt with inline text when systemPrompt is set", () => {
+    const cmd = agent.getLaunchCommand(makeLaunchConfig({ systemPrompt: "Be helpful" }));
+    expect(cmd).toContain("--system-prompt 'Be helpful'");
+  });
```
[the remark this answers is not in the document]
Comment (suraj-markup) [reply packages/plugins/agent-codex/src/index.test.ts:202]: Fixed in 0a5fea1 — updated all three test expectations to match the correct Codex CLI flags: `-c developer_instructions=...` for inline text and `-c model_instructions_file=...` for file-based prompts. All 85 codex plugin tests pass.
